src/components/Palette/Palette.tsx

```diff
@@ -0,0 +1,287 @@
+    };
+
+    const rows = React.useMemo(() => getRows(options, columns), [columns, options]);
+
+    const focusOnOptionWithIndex = React.useCallback((index: number) => {
+        if (!innerRef.current) return;
+
+        const $options: HTMLButtonElement[] = Array.from(
+            innerRef.current.querySelectorAll(`.${paletteClassNames.option()}`),
+        );
+
+        $options[index].focus();
+
+        setFocusedOptionIndex(index);
+    }, []);
+
+    const onKeyDown = React.useCallback(
+        (event: KeyboardEvent) => {
+            if (
+                focusedOptionIndex === undefined ||
+                focusedRow === undefined ||
+                focusedColumn === undefined
+            ) {
+                return;
+            }
+
+            let newIndex = focusedOptionIndex;
+
+            if (event.code === 'ArrowUp') {
+                event.preventDefault();
+                newIndex = focusedOptionIndex - columns;
+            } else if (event.code === 'ArrowRight') {
+                event.preventDefault();
+                newIndex = focusedOptionIndex + 1;
+            } else if (event.code === 'ArrowDown') {
+                event.preventDefault();
+                newIndex = focusedOptionIndex + columns;
+            } else if (event.code === 'ArrowLeft') {
+                event.preventDefault();
+                newIndex = focusedOptionIndex - 1;
+            }
+
+            if (newIndex === focusedOptionIndex || newIndex < 0 || newIndex >= options.length) {
+                return;
+            }
+
+            focusOnOptionWithIndex(newIndex);
+        },
+        [
+            focusedOptionIndex,
+            focusedRow,
+            focusedColumn,
+            options.length,
+            focusOnOptionWithIndex,
+            columns,
+        ],
+    );
+
+    React.useEffect(() => {
+        document.addEventListener('keydown', onKeyDown);
+        return () => document.removeEventListener('keydown', onKeyDown);
+    }, [onKeyDown]);
+
+    const onFocusGrid = React.useCallback(() => {
+        focusOnOptionWithIndex(0);
+    }, [focusOnOptionWithIndex]);
+
+    const onBlurGrid = React.useCallback(() => {
+        setFocusedOptionIndex(undefined);
+    }, []);
+
+    const onFocusOption = React.useCallback(
+        (event: React.FocusEvent<HTMLButtonElement>) => {
+            const newIndex = options.findIndex((option) => option.value === event.target.value);
+            if (newIndex >= 0) {
+                focusOnOptionWithIndex(newIndex);
+            }
+
+            onFocus?.(event);
+            event.stopPropagation();
+        },
+        [options, focusOnOptionWithIndex, onFocus],
+    );
+
+    const onBlurOption = React.useCallback(
+        (event: React.FocusEvent<HTMLButtonElement>) => {
+            onBlur?.(event);
+            setFocusedOptionIndex(undefined);
+            event.stopPropagation();
+        },
+        [onBlur],
+    );
+
+    return (
+        <div
```

Comment:
You can check w3c pattern [here](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) to clarify the roles of the elements

Reply:
I think I used the right roles except for the `checkbox` for the `Button`
I also added a wrapper for the `Button` with `gridcell` role (I'm not sure if it's necessary, but it's recommended in the article you sent if I understand correctly)
The `grid` and `row` roles seem to be right